src/BuildLoop.cpp

```diff
@@ -139,9 +119,67 @@ static bool AllDependenciesAreSuccesful(BuildQueue *queue, RuntimeNode *runtime_
     return true;
 }
 
+static bool AddNodeToWorkStackIfNotAlreadyPresent(BuildQueue* queue, RuntimeNode* runtime_node)
+{
+    int runtime_node_index = int(runtime_node - queue->m_Config.m_RuntimeNodes);
+    return BufferAppendOneIfNotPresent(&queue->m_WorkStack, queue->m_Config.m_Heap, runtime_node_index);
+}
+
+int EnqueueNodeWithoutWakingAwaiters(BuildQueue *queue, MemAllocLinear* scratch, RuntimeNode *runtime_node, RuntimeNode* queueing_node)
+{
+    if (RuntimeNodeHasEverBeenQueued(runtime_node))
+        return 0;
+
+    LogFirstTimeEnqueue(scratch, runtime_node, queueing_node);
+    queue->m_AmountOfNodesEverQueued++;
+    RuntimeNodeFlagQueued(runtime_node);
+
+    //enqueueing a node means that we know we need it to complete our build. Some nodes
+    //we know can be processed immediately:
+    //1) those whose dependencies have all been completed,
+    //2) those who are marked as leaf input cacheable && leaf input caching is turned on.
+    //
+    //we will only put nodes from these two categories on the workstack. Any other nodes
+    //we will only mark them as queued, but we don't actually put them on the workstack, since we already
+    //know they cannot yet immediately be acted upon. They will be put on the workstack when their dependencies finish.
+    int placed_on_workstack_count = 0;
+    if (AllDependenciesAreFinished(queue,runtime_node) || IsNodeCacheableByLeafInputsAndCachingEnabled(queue,runtime_node))
+    {
+        if (AddNodeToWorkStackIfNotAlreadyPresent(queue, runtime_node))
+            placed_on_workstack_count++;
+    } else {
+        //ok, so in this case the queued node is not immediately actionable. so we don't put it on the workstack, but instead queue up all our tobuild dependencies,
+        //so that our node becomes immediately actionable in the future. We don't blindly always do this, because in the case where this node is leaf input cacheable
+        //we might get a cache hit, and it won't be necissery at all to build any of the dependencies.
```

Comment:
nitpick: s/necissery/necessary/ (I'm bad at spelling that word too)

Reply:
fixed 8882e09